server/src/main/java/org/elasticsearch/search/retriever/RankDocsRetrieverBuilder.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.retriever;
+
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.DisMaxQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryRewriteContext;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.search.rank.RankDoc;
+import org.elasticsearch.search.retriever.rankdoc.RankDocsQueryBuilder;
+import org.elasticsearch.search.retriever.rankdoc.RankDocsSortBuilder;
+import org.elasticsearch.search.sort.ScoreSortBuilder;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+/**
+ * An {@link RetrieverBuilder} that is used to retrieve documents based on the rank of the documents.
+ */
+public class RankDocsRetrieverBuilder extends RetrieverBuilder {
```

Comment:
I don't see this used anywhere. I don't understand how it should be used. 

I would assume users would NEVER provide this via API, and this is only used as an intermediary for things to rewrite between?

But I see no evidence of this, and i am confusd.

Reply:
Yes, none of the query, sort, or retriever components added in this PR should be directly accessed via the API. Instead, they should be part of a higher-level retriever, such as a combined retriever (RRF). We discussed adding this combined retriever in the PR with Panos but decided to split it off since the changes are already significant.

Reply:
Yeap, as Jim mentioned this is an internal only API that for now remains idle in production (a PR on this will follow). There is an example of how this could be leveraged in RRF in the `RankDocRetrieverBuilderIT# CompoundRetrieverWithRankDocs`.